src/walberla_bridge/src/lattice_boltzmann/LBWalberlaImpl.hpp

```diff
@@ -389,8 +391,52 @@ class LBWalberlaImpl : public LBWalberlaBase {
     if (not lower_bc or not upper_bc) {
       return std::nullopt;
     }
-    assert(&(*(lower_bc->block)) == &(*(upper_bc->block)));
-    return {CellInterval(lower_bc->cell, upper_bc->cell)};
+    Cell const global_lower_cell = lower_bc->cell;
+    Cell const global_upper_cell =
+        Cell(static_cast<int>(upper_bc->cell[0] +
+                              upper_bc->block->getAABB().min()[0] -
+                              lower_bc->block->getAABB().min()[0]),
+             static_cast<int>(upper_bc->cell[1] +
+                              upper_bc->block->getAABB().min()[1] -
+                              lower_bc->block->getAABB().min()[1]),
+             static_cast<int>(upper_bc->cell[2] +
+                              upper_bc->block->getAABB().min()[2] -
+                              lower_bc->block->getAABB().min()[2]));
+    return {CellInterval(global_lower_cell, global_upper_cell)};
+  }
+
+  // Interval within local block
+  [[nodiscard]] std::optional<CellInterval> get_block_interval(
+      Utils::Vector3i const &lower_corner, Utils::Vector3i const &upper_corner,
+      Utils::Vector3i const &block_offset, IBlock const &block) const {
+    auto block_lower_corner = to_vector3i(block.getAABB().min());
+    if (upper_corner[0] < block_lower_corner[0] or
+        upper_corner[1] < block_lower_corner[1] or
+        upper_corner[2] < block_lower_corner[2]) {
+      return std::nullopt;
+    }
+    for (uint_t f = 0u; f < 3u; ++f) {
+      block_lower_corner[f] = std::max(block_lower_corner[f], lower_corner[f]);
+    }
+    auto block_upper_corner = to_vector3i(block.getAABB().max());
+    if (lower_corner[0] > block_upper_corner[0] or
+        lower_corner[1] > block_upper_corner[1] or
+        lower_corner[2] > block_upper_corner[2]) {
+      return std::nullopt;
+    }
+    for (uint_t f = 0u; f < 3u; ++f) {
+      block_upper_corner[f] = std::min(block_upper_corner[f], upper_corner[f]);
+    }
+    block_upper_corner -= Utils::Vector3i::broadcast(1);
+    Cell const block_lower_cell =
+        Cell(static_cast<int>(block_lower_corner[0] - block_offset[0]),
+             static_cast<int>(block_lower_corner[1] - block_offset[1]),
+             static_cast<int>(block_lower_corner[2] - block_offset[2]));
+    Cell const block_upper_cell =
+        Cell(static_cast<int>(block_upper_corner[0] - block_offset[0]),
+             static_cast<int>(block_upper_corner[1] - block_offset[1]),
+             static_cast<int>(block_upper_corner[2] - block_offset[2]));
+    return {CellInterval(block_lower_cell, block_upper_cell)};
```

Comment:
* `Cell const` -> `auto const` since the rhs already indicates the type.
* are the static casts necessary? the result of the operations seem to already be `int`
* would `return {{block_lower_cell, block_upper_cell}};` work here? I never remember the exact rule, but I think the outermost braces should invoke the optional constructor, and the innermost braces should invoke the cell interval constructor. For initializer lists and pointer-valued element different rules would apply.

Reply:
I have rewritten `Cell const` to `auto const` and removed static_cast. And I have removed Braces and wrote `return CellInterval(block_lower_cell, block_upper_cell);`